packages/angular_devkit/core/src/jobs/README.md

```diff
@@ -0,0 +1,464 @@
+the `JobInput` interface dictates what kind of message it is sending:
+
+1. `JobInputKind.Ping`. A simple message that should be answered with `JobEventKind.Pong` when
+    the job is responsive. The `id` field of the message should be used when returning `Pong`.
+1. `JobInputKind.Stop`. The job should be stopped. This is used when cancelling/unsubscribing from
+    the `output` (or by calling `stop()`). Any inputs or outputs after this event will be ignored.
+1. `JobInputKind.ChannelMessage`, `JobInputKind.ChannelError` and `JobInputKindChannelComplete` 
+    are used when sending inputs to a job. These correspond to the `next`, `error` and `complete`
+    methods of an `Observer` and are reported to the Job through its `context.input` Observable.
+
+Using the `createJob()` helper, all those events are automatically handled by the boilerplate 
+code. If you need direct access to raw inputs, you should subscribe to the `context.rawInput` 
+Observable.
+
+### Raw Output Stream
+The `Job<>` interface also contains a `rawOutput` member (of type `Observable<JobEvent<O>>` where
+`O` is the typed output of the job) which is the output complement of `rawInput`. All messages 
+sent _from_ the Job goes through this stream. The `kind` member of the `JobEvent<O>` interface
+dictates what kind of message it is sending:
+
+1. `JobEventKind.Create`. The `Job<>` was created, its dependencies are done, and the library is 
+    validating Argument and calling the internal Job code.
+1. `JobEventKind.Start`. The Job code itself should send that event when started. `createJob()`
+    will do it automatically.
+1. `JobEventKind.End`. The Job has ended. This is done by the Job itself and should always be
+    sent when completed. The scheduler will listen to this event to set the state and unblock
+    dependent jobs. `createJob()` automatically send this event.
+1. `JobEventKind.Pong`. The Job should answer a `JobInputKind.Ping` event with this.
+    Automatically done by `createJob()`.
+1. `JobEventKind.Progress`. The Job is progressing. This has a progress value which will be
+    capped between 0 and 1.
+1. `JobEventKind.Log`. A logging event (side effect that should be shown to the user).
+1. `JobEventKind.Output`. An `Output` has been generated by the Job.
+1. `JobInputKind.ChannelMessage`, `JobInputKind.ChannelError` and `JobInputKindChannelComplete`
+    are used for output channels. These correspond to the `next`, `error` and `complete` methods
+    of an `Observer` and are available to the callee through the `job.channels` map of Observable.
+
+Those events can be accessed directly through the `job.rawOutput` member. The Job itself should
+return an `Observable<JobEvent<O>>`. The `createJob()` helper handles most of use cases of this 
+and makes it easier for jobs to handle this.
+
+## Job Dispatchers
+Dispatchers are an additional way to synchronize and execute logic on scheduling the job, from the 
```

Comment:
Do dispatchers offer any way to synchronize jobs? It looks like they allow a conditional job but that any synchronization at this point is done via strategies.

Reply:
No. Dispatchers are literally "if X run job named Y, else run job named Z".

Example of a dispatcher without the Dispatcher class:

```typescript
import { jobs } from '@angular-devkit/core';
import { isYarnInstalled } from '...';

export default jobs.createJob(
  (arg, context) => {
    if (isYarnInstalled()) {
      context.scheduler('yarn-install', arg);
    } else {
      context.scheduler('npm-install', arg);
    }
  },
);
```

The `Dispatcher` class just makes it easier to manage that boilerplate, which can be repetitive.